diag_manager/fms_diag_object_container.F90

```diff
@@ -0,0 +1,228 @@
+!***********************************************************************
+!*                   GNU Lesser General Public License
+!*
+!* This file is part of the GFDL Flexible Modeling System (FMS).
+!*
+!* FMS is free software: you can redistribute it and/or modify it under
+!* the terms of the GNU Lesser General Public License as published by
+!* the Free Software Foundation, either version 3 of the License, or (at
+!* your option) any later version.
+!*
+!* FMS is distributed in the hope that it will be useful, but WITHOUT
+!* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+!* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+!* for more details.
+!*
+!* You should have received a copy of the GNU Lesser General Public
+!* License along with FMS.  If not, see <http://www.gnu.org/licenses/>.
+!***********************************************************************
+
+!> @defgroup fms_diag_object_container_mod fms_diag_object_container_mod
+!> @ingroup diag_manager
+!> @brief fms_diag_object_container_mod defines a container for inserting, removing and
+!! searching for <TT>fms_diag_object</TT> instances
+!!
+!> @author Miguel Zuniga Miguel.Zuniga@noaa.gov
+!!
+!! <TT>fms_diag_object_container_mod</TT> defines a container for inserting, removing and
+!! searching for <TT>fms_diag_object</TT> instances.
+!!
+!! This version uses the <TT>fms_doubly_linked_list_mod</TT> as the underlying
+!! implementation. Future vesions may change the underlying implementation but
+!! the user interface will not change.
+!!
+!> @file
+!> @brief File for @ref fms_diag_object_container_mod
+MODULE fms_diag_object_container_mod
+   use fms_diag_object_mod, only: fms_diag_object
+   USE fms_mod, ONLY: error_mesg, FATAL, WARNING, NOTE
+
+   !! Since this version is based on the FDS linked list:
+   use fms_doubly_linked_list_mod, only : linked_list_type, literator_type
+
+   implicit none
+
+   !> @brief A container of fms_diag_object instances providing insert, remove ,
+   !!  find/search, and size public member functions. Iterator is provided by
+   !!  the associated iterator class (see  dig_obj_iterator class).
+   !!
+   !!  This version does not enforce uniqueness of ID keys (I.e. it is not a set).
+   !!
+   type, public:: fms_diag_object_container_type
+      TYPE (linked_list_type), ALLOCATABLE :: the_linked_list !!This version based on the FDS linked_list.
+   contains
+      procedure :: insert => insert_diag_object
+      procedure :: remove => remove_diag_object
+      procedure :: find   => find_diag_object
+      procedure :: size   => get_num_objects
+      procedure :: iterator  => get_iterator
+      final :: destructor
+   end type fms_diag_object_container_type
+
+
+   !> \brief Iterator used to traverse the objects of the container.
+   type, public :: fms_diag_obj_iterator_type
+      type(literator_type) :: liter !!this version based on the FDS linked_list
```

Comment:
@thomas-robinson 
Concerning the ```this``` usage in places such as ```pure function something (this, var) result (res)```, are you shure its wise to put in a description?  Its all going to be the the same for each function : 
```!> The instance of the class that this function is bount to```.

Reply:
Yes because it's needed for the documentation web page generation.  Also, the `!<` needs to be the way I've typed it, not `!>` when documenting variables.  Use `!>` when making documentation comments.

Reply:
Done.
OK. I added documentation for this thought the code, and corrected the direction to ```!<``` for documenting variables. 
PS I just learned about the ```cmake -C docs```